src/ufo-reduce-task.c

```diff
@@ -85,23 +130,154 @@ ufo_reduce_task_process (UfoTask *task,
+
+    UFO_RESOURCES_CHECK_CLERR (clSetKernelArg (priv->kernel, 2, priv->local_size * sizeof (cl_float), NULL));
+
+    /* Balance the load and process multiple times until the global reduction
+     * result is stored in the first pixel. One work item in the kernel
+     * processes more pixels (global work size is thus less than the input
+     * size). At the same time, we try to have many groups in order to have good
+     * occupancy. */
+    num_groups = input_size;
+    while (num_groups > 1) {
+        real_size = num_groups;
+        /* Make sure global work size divides the local work size */
+        /* Number of groups processing *real_size* data */
+        num_groups = (real_size - 1) / priv->local_size + 1;
+        /* Balance the load, i.e. half of the work goes to work items, half to
+         * groups. This way the GPU is busy on the work item level (because
+         * every work item processes more input pixels) and occupancy is good
+         * because there are many groups. */
+        pixels_per_thread = (gint) (ceil (sqrt (num_groups)));
+        num_groups = (num_groups - 1) / pixels_per_thread + 1;
+        global_work_size = num_groups * priv->local_size;
+        if (!priv->result) {
+            priv->result = clCreateBuffer (priv->context,
+                                           CL_MEM_READ_WRITE,
+                                           num_groups * sizeof (cl_float),
+                                           NULL,
+                                           &error);
+            UFO_RESOURCES_CHECK_CLERR (error);
+            UFO_RESOURCES_CHECK_CLERR (clSetKernelArg (priv->kernel, 1, sizeof (cl_mem), &priv->result));
+        } else {
+            in_mem = priv->result;
         }
+
+        g_debug ("real size: %d global size: %lu padded n: %d G: %d PPT: %d",
+                 real_size, global_work_size, n, num_groups, pixels_per_thread);
+        UFO_RESOURCES_CHECK_CLERR (clSetKernelArg (priv->kernel, 0, sizeof (cl_mem), &in_mem));
+        UFO_RESOURCES_CHECK_CLERR (clSetKernelArg (priv->kernel, 3, sizeof (cl_int), &real_size));
+        UFO_RESOURCES_CHECK_CLERR (clSetKernelArg (priv->kernel, 4, sizeof (cl_int), &pixels_per_thread));
+        ufo_profiler_call (profiler, cmd_queue, priv->kernel, 1, &global_work_size, &priv->local_size);
+    }
+
+
+    /* Result is stored in the first pixel */
+    UFO_RESOURCES_CHECK_CLERR (clEnqueueReadBuffer (cmd_queue,
+                                                    priv->result,
+                                                    CL_TRUE,
+                                                    0,
+                                                    sizeof (cl_float),
+                                                    &result,
+                                                    0, NULL, NULL));
+    if (priv->mode == MODE_MEAN) {
+        result /= input_size;
     }
 
+    /* Pass the original data on intact */
+    ufo_buffer_copy (inputs[0], output);
```

Comment:
Could you check if the recently added `ufo_buffer_swap_data` call works here as well? That would avoid the copy.